ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala

```diff
@@ -116,33 +118,62 @@ object Main extends StrictLogging {
     logger.info(s"$config")
 
     val et: ET[ExitCode] = for {
-      ledgerId <- either(
-        getLedgerId(config.jwt).leftMap(_ =>
-          new IllegalArgumentException("Cannot infer Ledger ID from JWT")
-        )
-      ): ET[LedgerId]
 
       _ <- either(
         config.traverse(s => resolveSimulationClass(s))
       ): ET[Config[Class[_ <: Simulation]]]
 
       exitCode <- rightT(
-        main2(ledgerId, config)
+        main2(config)
       ): ET[ExitCode]
 
     } yield exitCode
 
     et.run
   }
 
-  private def main2(ledgerId: LedgerId, config: Config[String])(implicit
+  private def main2(config: Config[String])(implicit
       asys: ActorSystem,
       mat: Materializer,
       aesf: ExecutionSequencerFactory,
       ec: ExecutionContext,
       elg: EventLoopGroup,
-  ): Future[ExitCode] =
-    withLedger(config.dars, ledgerId.unwrap) { (ledgerPort, _, _) =>
+  ): Future[ExitCode] = {
+    // For custom tokens we need to extract the expected ledger id, however for
+    // user tokens we can just choose any ledger id, because it doesn't contain any.
+    // If we have a user token we also extract it here so we can allocate the user later
+    // after we started the ledger.
+    val (userIdOpt, ledgerId) =
+      HttpService
+        .decodeJwt(config.jwt)
+        .leftMap(_.message)
+        .flatMap { decodedJwt =>
+          CreateFromUserToken
+            .parseAndDecodeUserToken(decodedJwt)
+            .bimap(
+              error => s"This is neither a custom token nor a user token: ${error.message}",
+              {
+                case StandardJWTPayload(userId, _, _) =>
+                  (Some(userId), LedgerId("perf-runner"))
+                case CustomDamlJWTPayload(ledgerId, _, _, _, _, _, _) =>
+                  (
+                    None,
+                    LedgerId(
+                      ledgerId.getOrElse(throw new Exception("No ledger id given in custom token"))
```

Comment:
Rather not throw exceptions, but lift disjunction into future, and have a simple for comp on everything, failing fast.